File: src/test/java/imperative/parallel/ForkJoinConcat.java
package imperative.parallel;/*
 * Copyright (c) 2019 - Present, Gopal S Akshintala
 * This source code is licensed under the Creative Commons Attribution-ShareAlike 4.0 International License.
 * 	http://creativecommons.org/licenses/by-sa/4.0/
 */

import org.junit.jupiter.api.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.ForkJoinPool;
import java.util.concurrent.RecursiveTask;

import static common.Common.EXPECTED_RESULT;
import static common.Common.TEAM;
import static imperative.ImperativeLastName.concatLastNames;
import static imperative.parallel.Util.AVAILABLE_CORES;
import static imperative.parallel.Util.concatResults;
import static org.junit.jupiter.api.Assertions.assertEquals;

public class ForkJoinConcat {
    
    @Test
    void parallelWithForkJoinPool() {
        var forkJoinPool = new ForkJoinPool(AVAILABLE_CORES);
        var actualResult = forkJoinPool.invoke(new MyRecursiveTask(TEAM));
        System.out.println(actualResult);
        assertEquals(EXPECTED_RESULT, actualResult);
    }

    static class MyRecursiveTask extends RecursiveTask<String> {
        private static final long serialVersionUID = -5978274303314688527L;

        private static final int MIN_TEAM_SIZE = 2; // In real-world, DO NOT have it below 10,000
        private final List<String> team;

        MyRecursiveTask(List<String> team) {
            this.team = team;
        }

        @Override
        protected String compute() {
            if (team.size() > MIN_TEAM_SIZE) {
                var mid = team.size() / 2;
                final var myRecursiveTask1 = new MyRecursiveTask(team.subList(0, mid));
                final var myRecursiveTask2 = new MyRecursiveTask(team.subList(mid, team.size()));

                myRecursiveTask1.fork();
                myRecursiveTask2.fork();

                var results = new ArrayList<String>();
                results.add(myRecursiveTask1.join());
                results.add(myRecursiveTask2.join());
                return concatResults(results);
            } else {
                return concatLastNames(team);
            }
        }
    }

}

File: src/test/kotlin/kt/imperative/parallel/ForkJoinConcat.kt
/* gakshintala created on 4/14/20 */
package kt.imperative.parallel

import kt.common.EXPECTED_RESULT
import kt.common.TEAM
import kt.imperative.ImperativeConcat.Companion.concatLastNames
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test
import java.util.concurrent.ForkJoinPool
import java.util.concurrent.RecursiveTask

internal class MyRecursiveTask(private val team: List<String?>?) : RecursiveTask<String>() {
    override fun compute(): String = team?.let {
        if (it.size > MIN_TEAM_SIZE) {
            val mid = it.size / 2
            val myRecursiveTask1 = MyRecursiveTask(it.subList(0, mid))
            val myRecursiveTask2 = MyRecursiveTask(it.subList(mid, it.size))
            myRecursiveTask1.fork()
            myRecursiveTask2.fork()

            concatResultsFromForks(myRecursiveTask1, myRecursiveTask2)
        } else {
            concatLastNames(it) // Using concatLastNames as MIN_TEAM_SIZE can be greater than 2
        }
    } ?: ""

    companion object {
        private const val serialVersionUID = -5978274303314688527L
        private const val MIN_TEAM_SIZE = 2 // In real-world, DO NOT have it below 10,000
    }
}

class ForkJoinConcat {

    @Test
    fun parallelWithForkJoinPool() {
        val forkJoinPool = ForkJoinPool(AVAILABLE_CORES)
        val actualResult = forkJoinPool.invoke(MyRecursiveTask(TEAM))
        println(actualResult)
        Assertions.assertEquals(EXPECTED_RESULT, actualResult)
    }
}
